AtTools/AtEDistortionModel.cxx

```diff
@@ -0,0 +1,191 @@
+   }
+
+   LOG(debug) << " Input point : " << input.X() << " " << input.Y() << " " << input.Z() << "\n";
+
+   Double_t inputRad = TMath::Sqrt(TMath::Power(input.X(), 2) + TMath::Power(input.Y(), 2));
+   Double_t inputAng = TMath::ATan2(input.Y(), input.X());
+   Double_t inputZ = input.Z();
+
+   inputZ = inputZ < 0 ? 0 : inputZ;
+   inputZ = inputZ > 1000 ? 1000 : inputZ;
+
+   auto [zcorr, radcorr, tracorr] = GetCorrectionFactors((Int_t)inputRad, (Int_t)inputZ);
+
+   Double_t outputRad = TMath::Sqrt(TMath::Power(inputRad + radcorr, 2) + TMath::Power(tracorr, 2));
+   Double_t outputAng = inputAng + TMath::ATan2(tracorr, inputRad + radcorr);
+
+   XYZPoint output;
+   output.SetX(outputRad * TMath::Cos(outputAng));
+   output.SetY(outputRad * TMath::Sin(outputAng));
+   output.SetZ(input.Z() - zcorr);
+
+   LOG(debug) << " Output point : " << output.X() << " " << output.Y() << " " << output.Z() << "\n";
+
+   return output;
+}
+
+XYZPoint AtEDistortionModel::ApplySpaceCharge(const XYZPoint &input)
+{
+   XYZPoint p;
+   return p;
+}
+
+Bool_t
+AtEDistortionModel::SetCorrectionMaps(const std::string &zlut, const std::string &radlut, const std::string &tralut)
+{
+
+   fZLUT.open(zlut, std::ifstream::in);
+   fRadLUT.open(radlut, std::ifstream::in);
+   fTraLUT.open(tralut, std::ifstream::in);
+
+   if (!fZLUT.fail() && !fRadLUT.fail() && !fTraLUT.fail()) {
+
+      SetBufferMap(fZLUT, fZMap.get());
+      SetBufferMap(fRadLUT, fRadMap.get());
+      SetBufferMap(fTraLUT, fTraMap.get());
+
+      SetGlobalMap(fZMap.get(), fRadMap.get(), fTraMap.get());
+
+      return kTRUE;
+
+   } else {
+
+      std::cerr << " AtEDistortionModel::SetCorrectionMaps - Error! Missing LUT "
+                << "\n";
+      std::cerr << zlut << "\n";
+      std::cerr << radlut << "\n";
+      std::cerr << tralut << "\n";
+      return kFALSE;
+   }
+
+   return kTRUE;
+}
+
+std::tuple<Double_t, Double_t, Double_t>
+AtEDistortionModel::GetCorrectionFactors(const Int_t &radius, const Int_t &zpos)
+{
+
+   if (fDistortionMap->size() == 0) {
+      LOG(error) << " Warning! Correction map is empty. "
+                 << "\n";
+      return std::make_tuple(0, 0, 0);
+   }
+
+   EFieldMapRef ref(radius, zpos);
+   if (fDistortionMap->find(ref) == fDistortionMap->end()) {
+      LOG(error) << " Correction factor not found. "
+                 << "\n";
+      return std::make_tuple(0, 0, 0);
+   } else {
+      auto corr = fDistortionMap->at(ref);
+      return std::make_tuple(corr.GetZCorr(), corr.GetRadCorr(), corr.GetTraCorr());
+   }
+}
+
+void AtEDistortionModel::SetBufferMap(std::ifstream &lut, BufferMap *map)
+{
+
+   while (!lut.eof()) {
+
+      std::string line;
+      std::getline(lut, line, '\r');
+      Int_t radius;
+      Double_t value;
+      std::istringstream buffer(line);
+      Int_t cnt = 0;
+
+      buffer >> radius;
+
+      while (buffer >> value) {
+
+         EFieldMapRef ref(radius, cnt);
+         map->emplace(ref, value);
```

Comment:
Could probably replace with `map->emplace({radius, cnt}, value);` Little shorted and imo just as readable.

Reply:
Wasnt able to compile it

Reply:
For this to compiled the suggested changes to the struct had to be made to so it could use aggregate initialization https://en.cppreference.com/w/cpp/language/aggregate_initialization

Reply:
Thats what I actually did, but it didnt like it
141:36: error: no matching function for call to ‘std::unordered_map<EFieldMapRef, double>::emplace(<brace-enclosed initializer list>, Double_t&)’
  141 |     map->emplace({radius,cnt},value);

Reply:
Someday I'll figure out why that is. I expected it to work like this `fAtMapPtr->AddAuxPad({10, 0, 0, 0}, "MCP_US");` but something must be different with the map emplace function signature. My guess not testing is emplace takes an r value, and it can't bind to whatever is actually being returned by the `{}` while in a the AddAuxPad it can bind to the const reference used there.